### ISSUE TITLE
Explicitly-written type on for-loop variable

### ISSUE THREAD

**Issue by mernst** [kelloggm/checker-framework#135]
When a programmer writes a type on a for loop variable, updates/assignments to the variable do not seem to be checked.  Consider this test case:

```java
import org.checkerframework.checker.index.qual.IndexFor;

public class ForIndexFor {

    public static void main(String[] args) {

        int[] ar = new int[4];
        //:: error: (?)
        for (@IndexFor("ar") int i = 0; i <= 5; i++) {
            ar[i] = i;
        }

        int[] ar2 = new int[4];
        for (int i = 0; i <= 5; i++) {
            //:: error: (array.access.unsafe.high)
            ar2[i] = i;
        }
    }
}
```

The Index Checker only issues a warning about the second loop, not the first one:

```
$ch/bin/javac -g ForIndexFor.java -processor index
ForIndexFor.java:15: error: [array.access.unsafe.high] Potentially unsafe array access: the index could be larger than the array's bound
            ar2[i] = i;
                ^
  found   : @UpperBoundUnknown int
  required: an integer less than ar2's length (@LTLengthOf("ar2"))
1 error
```

**Comment by mernst** [comment]
Additional test case, from duplicate report #142:

```java
import org.checkerframework.checker.index.qual.*;

class IndexTest {
  public static void main(String[] args) {
    String[] name = {"a", "b", "c", "d"};

    //:: error: (compound.assignment.type.incompatible)
    for (@LTOMLengthOf("name") int i = 0; i < name.length; ++i) {
      name[i] = name[i + 1];
    }

    //:: error: (compound.assignment.type.incompatible)
    for (@LTOMLengthOf("name") int i = 0; i < name.length; i++) {
      name[i] = name[i + 1];
    }
  }
}
```

**Comment by kanha95** [comment]
@mernst  I want to clarify one thing professor.

Consider this variant of the 1st for loop of your 1st test case:

   int[] ar2 = new int[4];
        for (int i = 0; i <2; i++) {
            //:: error: (array.access.unsafe.high)
            ar2[i] = i;
        }

Output:
 error: [array.access.unsafe.high] Potentially unsafe array access: the index could be larger than the array's bound
            System.out.println(i[j]);
  found   : @UpperBoundUnknown int
  required: an integer less than i's length (@LTLengthOf("i"))

It still raises an error. It really doesn't matter what range 'i' is operating between. It raises an error in all scenarios.

**Comment by mernst** [comment]
Yes, this is a false positive warning.  Does it come up in real-world code?

I am sure that you can make up artificial code that stymies the type-checker!  You have done so here (nice job).  But we'd like to focus our energy on situations that will affect users.  Thanks.

**Comment by kanha95** [comment]
@mernst  Although i didn't intend to do so! Actually i caught this error in one of my programs which i didn't reveal.

Here it is:

``
import java.util.Arrays;
import java.util.Scanner;

class IndexTest{
    
    public static void main(String... args){
      
        Scanner sc=new Scanner(System.in);
                        
        Integer[] ar1=new Integer[5];
        Integer[] ar2=new Integer[5];
        
        for (int i = 0; i < 5; i++) {
            ar1[i]=i;
        }
        ar2=Arrays.copyOf(ar1, 8);
        
       //length of array ar2 now changed to 8
        System.out.println(ar2.length);
        
        // raises error with index checker
        for (int i = 0; i < 6; i++) {
            System.out.println(ar2[i]);
        }
    }   
} 

``

Output without checker:
8
0
1
2
3
4
null

Now i really don't know if that would affect users . I just thought to let you know.

**Comment by mernst** [comment]
Your test case is valid.  It's a duplicate of issue #118 and issue #14.

Let's keep each issue in the issue tracker focused on one topic.  This one is about pre- versus post-increment.

**Comment by smillst** [comment]
This is probably related to issue #867, dataflow bug about post-increments.
